File: src/domain/preferences.ts
import {
  loadBundledIngredientCatalog,
  type IngredientCatalog,
} from "@/data/ingredientCatalog";
import {
  rankGeneratedCandidates,
  type CandidateScore,
  type RankGeneratedCandidateOptions,
  type ScoringContext,
  type ScoringGoal,
} from "@/domain/candidateScoring";
import type { RecipeMatch, RecipeMatchKind } from "@/domain/recipeMatcher";
import type { Recipe } from "@/domain/recipes";
import {
  generateSmoothies,
  type GeneratedSmoothie,
  type SmoothieGenerationOptions,
} from "@/domain/smoothieGenerator";

export const FEEDBACK_VALUES = [
  "liked",
  "neutral",
  "do_not_suggest",
] as const;

export type FeedbackValue = (typeof FEEDBACK_VALUES)[number];

export interface UserPreferences {
  readonly favoriteIngredientIds: readonly string[];
  readonly favoriteRecipeKeys: readonly string[];
  readonly excludedIngredientIds: readonly string[];
  readonly allergies: readonly string[];
  readonly vegan: boolean;
  readonly vegetarian: boolean;
  readonly dairyFree: boolean;
  readonly desiredSweetness: number | null;
  readonly desiredCreaminess: number | null;
  readonly refreshing: boolean;
  readonly filling: boolean;
  readonly proteinRich: boolean;
  readonly lowerCalorie: boolean;
  readonly breakfast: boolean;
  readonly postWorkout: boolean;
  readonly feedback: Readonly<Record<string, FeedbackValue>>;
}

export const DEFAULT_USER_PREFERENCES: UserPreferences = Object.freeze({
  favoriteIngredientIds: Object.freeze([]),
  favoriteRecipeKeys: Object.freeze([]),
  excludedIngredientIds: Object.freeze([]),
  allergies: Object.freeze([]),
  vegan: false,
  vegetarian: false,
  dairyFree: false,
  desiredSweetness: null,
  desiredCreaminess: null,
  refreshing: false,
  filling: false,
  proteinRich: false,
  lowerCalorie: false,
  breakfast: false,
  postWorkout: false,
  feedback: Object.freeze({}),
});

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function uniqueStrings(value: unknown): readonly string[] {
  if (!Array.isArray(value)) {
    return Object.freeze([]);
  }

  return Object.freeze([
    ...new Set(
      value
        .filter((item): item is string => typeof item === "string")
        .map((item) => item.trim())
        .filter(Boolean),
    ),
  ]);
}

function canonicalIngredientIds(
  value: unknown,
  catalog: IngredientCatalog,
): readonly string[] {
  const result: string[] = [];

  for (const item of uniqueStrings(value)) {
    const id = catalog.resolveId(item);
    if (id && !result.includes(id)) {
      result.push(id);
    }
  }

  return Object.freeze(result);
}

function preferenceScale(value: unknown): number | null {
  return typeof value === "number" &&
    Number.isInteger(value) &&
    value >= 0 &&
    value <= 5
    ? value
    : null;
}

function sanitizeFeedback(value: unknown): Readonly<Record<string, FeedbackValue>> {
  if (!isRecord(value)) {
    return Object.freeze({});
  }

  const result: Record<string, FeedbackValue> = {};
  for (const [key, rawValue] of Object.entries(value)) {
    if (
      key.trim() &&
      typeof rawValue === "string" &&
      FEEDBACK_VALUES.includes(rawValue as FeedbackValue)
    ) {
      result[key] = rawValue as FeedbackValue;
    }
  }

  return Object.freeze(result);
}

export function sanitizeUserPreferences(
  value: unknown,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): UserPreferences {
  if (!isRecord(value)) {
    return DEFAULT_USER_PREFERENCES;
  }

  return Object.freeze({
    favoriteIngredientIds: canonicalIngredientIds(
      value.favoriteIngredientIds ?? value.favorite_ingredients,
      catalog,
    ),
    favoriteRecipeKeys: uniqueStrings(
      value.favoriteRecipeKeys ?? value.favorite_recipes,
    ),
    excludedIngredientIds: canonicalIngredientIds(
      value.excludedIngredientIds ?? value.excluded_ingredients,
      catalog,
    ),
    allergies: uniqueStrings(value.allergies),
    vegan: value.vegan === true,
    vegetarian: value.vegetarian === true,
    dairyFree: value.dairyFree === true || value.dairy_free === true,
    desiredSweetness: preferenceScale(
      value.desiredSweetness ?? value.desired_sweetness,
    ),
    desiredCreaminess: preferenceScale(
      value.desiredCreaminess ?? value.desired_creaminess,
    ),
    refreshing: value.refreshing === true,
    filling: value.filling === true,
    proteinRich: value.proteinRich === true || value.protein_rich === true,
    lowerCalorie:
      value.lowerCalorie === true || value.lower_calorie === true,
    breakfast: value.breakfast === true,
    postWorkout: value.postWorkout === true || value.post_workout === true,
    feedback: sanitizeFeedback(value.feedback),
  });
}

export function generatedFeedbackKey(candidate: GeneratedSmoothie): string {
  return `generated:${[...candidate.ingredientIds].sort().join(",")}`;
}

export function storedFeedbackKey(recipe: Recipe | string): string {
  return `stored:${typeof recipe === "string" ? recipe : recipe.id}`;
}

export function ingredientAllowedByPreferences(
  ingredientId: string,
  preferences: UserPreferences,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): boolean {
  const ingredient = catalog.require(ingredientId);

  if (preferences.excludedIngredientIds.includes(ingredient.id)) {
    return false;
  }
  if (preferences.vegan && !ingredient.vegan) {
    return false;
  }
  if (preferences.dairyFree && ingredient.allergens.includes("milk")) {
    return false;
  }
  if (
    ingredient.allergens.some((allergen) =>
      preferences.allergies.includes(allergen),
    )
  ) {
    return false;
  }

  return true;
}

export function filterPantryByPreferences(
  pantryIds: readonly string[],
  preferences: UserPreferences,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): readonly string[] {
  const result: string[] = [];

  for (const value of pantryIds) {
    const ingredient = catalog.resolve(value);
    if (
      ingredient &&
      !result.includes(ingredient.id) &&
      ingredientAllowedByPreferences(ingredient.id, preferences, catalog)
    ) {
      result.push(ingredient.id);
    }
  }

  return Object.freeze(result);
}

export function candidateAllowedByPreferences(
  candidate: GeneratedSmoothie,
  preferences: UserPreferences,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): boolean {
  if (
    preferences.feedback[generatedFeedbackKey(candidate)] === "do_not_suggest"
  ) {
    return false;
  }

  return candidate.ingredientIds.every((ingredientId) =>
    ingredientAllowedByPreferences(ingredientId, preferences, catalog),
  );
}

export function recipeAllowedByPreferences(
  recipe: Recipe,
  preferences: UserPreferences,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): boolean {
  if (preferences.feedback[storedFeedbackKey(recipe)] === "do_not_suggest") {
    return false;
  }

  return recipe.required.every((item) =>
    ingredientAllowedByPreferences(item.ingredientId, preferences, catalog),
  );
}

export function scoringContextFromPreferences(
  preferences: UserPreferences,
  pantryIds: readonly string[],
): ScoringContext {
  const goals: ScoringGoal[] = [];

  if (preferences.refreshing) goals.push("refreshing");
  if (preferences.filling) goals.push("filling");
  if (preferences.proteinRich) goals.push("protein_rich");
  if (preferences.lowerCalorie) goals.push("lower_calorie");
  if (preferences.breakfast) goals.push("breakfast");
  if (preferences.postWorkout) goals.push("post_workout");

  const likedCandidateKeys = Object.entries(preferences.feedback)
    .filter(
      ([key, value]) => key.startsWith("generated:") && value === "liked",
    )
    .map(([key]) => key);

  for (const key of preferences.favoriteRecipeKeys) {
    if (key.startsWith("generated:") && !likedCandidateKeys.includes(key)) {
      likedCandidateKeys.push(key);
    }
  }

  return Object.freeze({
    pantryIngredientIds: Object.freeze([...pantryIds]),
    preferredIngredientIds: preferences.favoriteIngredientIds,
    desiredSweetness: preferences.desiredSweetness ?? undefined,
    desiredCreaminess: preferences.desiredCreaminess ?? undefined,
    goals: Object.freeze(goals),
    likedCandidateKeys: Object.freeze(likedCandidateKeys),
  });
}

export function withFeedback(
  preferences: UserPreferences,
  recipeKey: string,
  value: FeedbackValue,
): UserPreferences {
  if (!recipeKey.trim()) {
    throw new Error("recipeKey must not be empty");
  }

  return Object.freeze({
    ...preferences,
    feedback: Object.freeze({
      ...preferences.feedback,
      [recipeKey]: value,
    }),
  });
}

export function clearPreferenceFeedback(
  preferences: UserPreferences,
): UserPreferences {
  return Object.freeze({
    ...preferences,
    feedback: Object.freeze({}),
  });
}

export function toggleFavoriteIngredient(
  preferences: UserPreferences,
  ingredientId: string,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): UserPreferences {
  const canonical = catalog.resolveId(ingredientId);
  if (!canonical) {
    return preferences;
  }

  const active = preferences.favoriteIngredientIds.includes(canonical);
  return Object.freeze({
    ...preferences,
    favoriteIngredientIds: Object.freeze(
      active
        ? preferences.favoriteIngredientIds.filter((id) => id !== canonical)
        : [...preferences.favoriteIngredientIds, canonical],
    ),
  });
}

export function toggleFavoriteRecipe(
  preferences: UserPreferences,
  recipeKey: string,
): UserPreferences {
  if (!recipeKey.trim()) {
    return preferences;
  }

  const active = preferences.favoriteRecipeKeys.includes(recipeKey);
  return Object.freeze({
    ...preferences,
    favoriteRecipeKeys: Object.freeze(
      active
        ? preferences.favoriteRecipeKeys.filter((key) => key !== recipeKey)
        : [...preferences.favoriteRecipeKeys, recipeKey],
    ),
  });
}

function excludedIdsForGeneration(
  preferences: UserPreferences,
  catalog: IngredientCatalog,
): readonly string[] {
  const ids = new Set(preferences.excludedIngredientIds);

  if (preferences.dairyFree) {
    for (const ingredient of catalog.all()) {
      if (ingredient.allergens.includes("milk")) {
        ids.add(ingredient.id);
      }
    }
  }

  return Object.freeze([...ids]);
}

export function generateSmoothiesWithPreferences(
  pantryIds: readonly string[],
  preferences: UserPreferences,
  options: SmoothieGenerationOptions = {},
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): readonly GeneratedSmoothie[] {
  const requestedCount = Math.max(0, Math.floor(options.count ?? 5));
  if (requestedCount === 0) {
    return Object.freeze([]);
  }

  const filteredPantry = filterPantryByPreferences(
    pantryIds,
    preferences,
    catalog,
  );
  const candidates = generateSmoothies(
    filteredPantry,
    {
      ...options,
      count: Math.max(requestedCount, requestedCount * 3),
      vegan: preferences.vegan || options.vegan === true,
      excludedAllergens: Object.freeze([
        ...new Set([
          ...(options.excludedAllergens ?? []),
          ...preferences.allergies,
        ]),
      ]),
      excludedIngredientIds: Object.freeze([
        ...new Set([
          ...(options.excludedIngredientIds ?? []),
          ...excludedIdsForGeneration(preferences, catalog),
        ]),
      ]),
    },
    catalog,
  );

  return Object.freeze(
    candidates
      .filter((candidate) =>
        candidateAllowedByPreferences(candidate, preferences, catalog),
      )
      .slice(0, requestedCount),
  );
}

export function rankGeneratedCandidatesWithPreferences(
  candidates: readonly GeneratedSmoothie[],
  preferences: UserPreferences,
  pantryIds: readonly string[],
  options: RankGeneratedCandidateOptions = {},
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): readonly CandidateScore[] {
  const allowed = candidates.filter((candidate) =>
    candidateAllowedByPreferences(candidate, preferences, catalog),
  );

  return rankGeneratedCandidates(
    allowed,
    scoringContextFromPreferences(preferences, pantryIds),
    options,
    catalog,
  );
}

const MATCH_KIND_RANK: Readonly<Record<RecipeMatchKind, number>> = {
  exact: 3,
  substitution: 2,
  partial: 1,
};

function storedPreferenceBonus(
  match: RecipeMatch,
  preferences: UserPreferences,
): number {
  const key = storedFeedbackKey(match.recipe);
  let bonus = 0;

  if (preferences.favoriteRecipeKeys.includes(key)) {
    bonus += 0.08;
  }
  if (preferences.feedback[key] === "liked") {
    bonus += 0.08;
  }

  const requiredIds = new Set(
    match.recipe.required.map((item) => item.ingredientId),
  );
  if (preferences.favoriteIngredientIds.length > 0 && requiredIds.size > 0) {
    const favoriteHits = preferences.favoriteIngredientIds.filter((id) =>
      requiredIds.has(id),
    ).length;
    bonus +=
      (favoriteHits / requiredIds.size) * 0.05;
  }

  const tags = new Set(match.recipe.tags);
  if (
    preferences.breakfast &&
    tags.has("frühstück")
  ) {
    bonus += 0.03;
  }
  if (preferences.proteinRich && tags.has("protein")) {
    bonus += 0.03;
  }
  if (
    preferences.refreshing &&
    (tags.has("frisch") || tags.has("erfrischend"))
  ) {
    bonus += 0.03;
  }
  if (preferences.filling && tags.has("sättigend")) {
    bonus += 0.03;
  }

  return Math.min(0.2, bonus);
}

export function rankStoredMatchesWithPreferences(
  matches: readonly RecipeMatch[],
  preferences: UserPreferences,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): readonly RecipeMatch[] {
  return Object.freeze(
    matches
      .filter((match) =>
        recipeAllowedByPreferences(match.recipe, preferences, catalog),
      )
      .sort((left, right) => {
        const kindDifference =
          MATCH_KIND_RANK[right.kind] - MATCH_KIND_RANK[left.kind];
        if (kindDifference !== 0) {
          return kindDifference;
        }

        const scoreDifference =
          right.availabilityScore +
          storedPreferenceBonus(right, preferences) -
          (left.availabilityScore +
            storedPreferenceBonus(left, preferences));
        if (scoreDifference !== 0) {
          return scoreDifference;
        }

        const missingDifference =
          left.missingRequired.length - right.missingRequired.length;
        if (missingDifference !== 0) {
          return missingDifference;
        }

        return left.recipe.id < right.recipe.id
          ? -1
          : left.recipe.id > right.recipe.id
            ? 1
            : 0;
      }),
  );
}

File: smoothie/preferences.py
"""Local user preferences, restrictions, favorites and recipe feedback."""
from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum
import json
import os
from pathlib import Path
from typing import Any

from .generator import GeneratedSmoothie
from .ingredient_catalog import IngredientCatalog
from .matcher import RecipeMatch
from .recipes import Recipe
from .scoring import ScoringContext

_PROJECT_DATA_DIR = Path(__file__).resolve().parents[1] / "data"
_STATE_DIR = Path(os.environ.get("SMOOTHIE_STATE_DIR", _PROJECT_DATA_DIR))
DEFAULT_PREFERENCES_PATH = _STATE_DIR / "user_preferences.local.json"


class FeedbackValue(StrEnum):
    LIKED = "liked"
    NEUTRAL = "neutral"
    DO_NOT_SUGGEST = "do_not_suggest"


@dataclass
class UserPreferences:
    favorite_ingredients: set[str] = field(default_factory=set)
    favorite_recipes: set[str] = field(default_factory=set)
    excluded_ingredients: set[str] = field(default_factory=set)
    allergies: set[str] = field(default_factory=set)
    vegan: bool = False
    vegetarian: bool = False
    dairy_free: bool = False
    desired_sweetness: int | None = None
    desired_creaminess: int | None = None
    refreshing: bool = False
    filling: bool = False
    protein_rich: bool = False
    lower_calorie: bool = False
    breakfast: bool = False
    post_workout: bool = False
    feedback: dict[str, FeedbackValue] = field(default_factory=dict)

    def __post_init__(self) -> None:
        for name in ("desired_sweetness", "desired_creaminess"):
            value = getattr(self, name)
            if value is not None and (
                not isinstance(value, int) or isinstance(value, bool) or not 0 <= value <= 5
            ):
                raise ValueError(f"{name} must be None or an integer from 0 to 5")
        self.feedback = {
            key: value if isinstance(value, FeedbackValue) else FeedbackValue(value)
            for key, value in self.feedback.items()
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "UserPreferences":
        return cls(
            favorite_ingredients=set(data.get("favorite_ingredients", [])),
            favorite_recipes=set(data.get("favorite_recipes", [])),
            excluded_ingredients=set(data.get("excluded_ingredients", [])),
            allergies=set(data.get("allergies", [])),
            vegan=bool(data.get("vegan", False)),
            vegetarian=bool(data.get("vegetarian", False)),
            dairy_free=bool(data.get("dairy_free", False)),
            desired_sweetness=data.get("desired_sweetness"),
            desired_creaminess=data.get("desired_creaminess"),
            refreshing=bool(data.get("refreshing", False)),
            filling=bool(data.get("filling", False)),
            protein_rich=bool(data.get("protein_rich", False)),
            lower_calorie=bool(data.get("lower_calorie", False)),
            breakfast=bool(data.get("breakfast", False)),
            post_workout=bool(data.get("post_workout", False)),
            feedback=dict(data.get("feedback", {})),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "favorite_ingredients": sorted(self.favorite_ingredients),
            "favorite_recipes": sorted(self.favorite_recipes),
            "excluded_ingredients": sorted(self.excluded_ingredients),
            "allergies": sorted(self.allergies),
            "vegan": self.vegan,
            "vegetarian": self.vegetarian,
            "dairy_free": self.dairy_free,
            "desired_sweetness": self.desired_sweetness,
            "desired_creaminess": self.desired_creaminess,
            "refreshing": self.refreshing,
            "filling": self.filling,
            "protein_rich": self.protein_rich,
            "lower_calorie": self.lower_calorie,
            "breakfast": self.breakfast,
            "post_workout": self.post_workout,
            "feedback": {key: value.value for key, value in sorted(self.feedback.items())},
        }


class PreferenceStore:
    """Persist a single private profile as local JSON."""

    def __init__(self, path: Path | str = DEFAULT_PREFERENCES_PATH) -> None:
        self.path = Path(path)

    def load(self) -> UserPreferences:
        if not self.path.exists():
            return UserPreferences()
        try:
            payload = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            raise ValueError(f"Could not load preferences from {self.path}: {exc}") from exc
        if not isinstance(payload, dict):
            raise ValueError("Preference data root must be a JSON object")
        return UserPreferences.from_dict(payload)

    def save(self, preferences: UserPreferences) -> None:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        temporary = self.path.with_suffix(self.path.suffix + ".tmp")
        temporary.write_text(
            json.dumps(preferences.to_dict(), ensure_ascii=False, indent=2) + "\n",
            encoding="utf-8",
        )
        temporary.replace(self.path)

    def reset(self) -> None:
        try:
            self.path.unlink()
        except FileNotFoundError:
            pass


def generated_feedback_key(candidate: GeneratedSmoothie) -> str:
    return "generated:" + ",".join(sorted(candidate.ingredient_ids))


def stored_feedback_key(recipe: Recipe | str) -> str:
    recipe_id = recipe if isinstance(recipe, str) else recipe.id
    return f"stored:{recipe_id}"


def ingredient_allowed(
    ingredient_id: str,
    preferences: UserPreferences,
    catalog: IngredientCatalog,
) -> bool:
    ingredient = catalog.require(ingredient_id)
    if ingredient_id in preferences.excluded_ingredients:
        return False
    if preferences.vegan and not ingredient.vegan:
        return False
    if preferences.dairy_free and "milk" in ingredient.allergens:
        return False
    if preferences.allergies.intersection(ingredient.allergens):
        return False
    return True


def filter_pantry(
    pantry_ids: list[str] | tuple[str, ...] | set[str] | frozenset[str],
    preferences: UserPreferences,
    catalog: IngredientCatalog,
) -> list[str]:
    return [
        ingredient_id
        for ingredient_id in dict.fromkeys(pantry_ids)
        if catalog.resolve(ingredient_id) is not None
        and ingredient_allowed(ingredient_id, preferences, catalog)
    ]


def recipe_allowed(
    recipe: Recipe,
    preferences: UserPreferences,
    catalog: IngredientCatalog,
) -> bool:
    if preferences.feedback.get(stored_feedback_key(recipe)) == FeedbackValue.DO_NOT_SUGGEST:
        return False
    return all(
        ingredient_allowed(item.ingredient_id, preferences, catalog)
        for item in recipe.required
    )


def candidate_allowed(candidate: GeneratedSmoothie, preferences: UserPreferences) -> bool:
    return (
        preferences.feedback.get(generated_feedback_key(candidate))
        != FeedbackValue.DO_NOT_SUGGEST
    )


def scoring_context_from_preferences(
    preferences: UserPreferences,
    pantry_ids: list[str] | tuple[str, ...] | set[str] | frozenset[str],
) -> ScoringContext:
    goals = {
        name
        for name in (
            "refreshing",
            "filling",
            "protein_rich",
            "lower_calorie",
            "breakfast",
            "post_workout",
        )
        if getattr(preferences, name)
    }
    liked = {
        key
        for key, value in preferences.feedback.items()
        if value == FeedbackValue.LIKED and key.startswith("generated:")
    }
    liked.update(
        key for key in preferences.favorite_recipes if key.startswith("generated:")
    )
    return ScoringContext(
        pantry_ids=frozenset(pantry_ids),
        preferred_ids=frozenset(preferences.favorite_ingredients),
        desired_sweetness=preferences.desired_sweetness,
        desired_creaminess=preferences.desired_creaminess,
        goals=frozenset(goals),
        liked_candidate_keys=frozenset(liked),
    )


def set_feedback(
    preferences: UserPreferences,
    recipe_key: str,
    value: FeedbackValue,
) -> None:
    if not recipe_key:
        raise ValueError("recipe_key must not be empty")
    preferences.feedback[recipe_key] = value


def clear_feedback(preferences: UserPreferences) -> None:
    preferences.feedback.clear()


def rank_stored_matches_with_preferences(
    matches: list[RecipeMatch],
    preferences: UserPreferences,
) -> list[RecipeMatch]:
    """Apply small, transparent soft boosts without overriding pantry availability."""

    def bonus(match: RecipeMatch) -> float:
        key = stored_feedback_key(match.recipe)
        value = 0.0
        if key in preferences.favorite_recipes:
            value += 0.08
        if preferences.feedback.get(key) == FeedbackValue.LIKED:
            value += 0.08
        required = {item.ingredient_id for item in match.recipe.required}
        if preferences.favorite_ingredients and required:
            value += (
                len(required & preferences.favorite_ingredients) / len(required)
            ) * 0.05
        goal_tags = {
            "breakfast": "frühstück",
            "protein_rich": "protein",
            "refreshing": "frisch",
            "filling": "sättigend",
        }
        for goal, tag in goal_tags.items():
            if getattr(preferences, goal) and tag in match.recipe.tags:
                value += 0.03
        return min(value, 0.20)

    return sorted(
        matches,
        key=lambda match: (
            not match.exact,
            -(match.score + bonus(match)),
            len(match.missing_required),
            match.recipe.id,
        ),
    )
